src/python/pants/engine/internals/synthetic_targets.py

```diff
@@ -0,0 +1,132 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+"""Synthetic targets is a concept of injecting targets into the build graph that doesn't have a home
+in any BUILD file.
+
+This is achieved by declaring a union member for the `SyntheticTargetAdaptorsRequest` union along
+with a rule taking that member and returning a `SyntheticTargetAdaptors`.
+
+Example demonstarting how to register synthetic targets:
+
+    from pants.engine.internals.synthetic_targets import (
+        SyntheticTargetAdaptor,
+        SyntheticTargetAdaptors,
+        SyntheticTargetAdaptorsRequest,
+    )
+    from pants.engine.unions import UnionRule
+    from pants.engine.rules import collect_rules, rule
+
+
+    class SyntheticExampleTargetAdaptorsRequest(SyntheticTargetAdaptorsRequest):
+        pass
+
+
+    @rule
+    async def example_synthetic_targets(_: SyntheticExampleTargetAdaptorsRequest) -> SyntheticTargetAdaptors:
+        return SyntheticTargetAdaptors(
+            [
+                SyntheticTargetAdaptor.create(
+                    "<spec-path>",
+                    "<target-alias>",
+                    "<target-name>",
+                    **target_kwargs,
+                ),
+                ...
+            ]
+        )
+
+
+    def rules():
+        return (
+            *collect_rules(),
+            UnionRule(SyntheticTargetAdaptorsRequest, SyntheticExampleTargetAdaptorsRequest),
+            ...
+        )
+"""
+from __future__ import annotations
+
+import itertools
+import os.path
+from dataclasses import dataclass
+from typing import cast
+
+from pants.engine.collection import Collection
+from pants.engine.internals.mapper import AddressMap
+from pants.engine.internals.target_adaptor import TargetAdaptor
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.unions import UnionMembership, union
+
+SYNTHETIC_BUILD_FILE_NAME = "-BUILD-synthetic-"
```

Comment:
Not super happy with this one. Want one that clearly says it's made up, but at the same time reflects the notion of a BUILD file, but not.

I don't believe this will ever show up, as it is used when presenting target name conflicts between BUILD files in the same path, but in this case, that will either merge or override so should never be an issue.